File: rightsmitra-backend/routes/cases.js
const store = require('../lib/store');

// GET /api/cases?owner_id=guest_xxx or user_xxx
function listCases(req, res, body, query) {
  const owner_id = query.owner_id || req.user?.user_id;
  if (!owner_id) return res.status(400).json({ error: 'owner_id query param required (guest_id or user_id)' });

  const cases = store
    .filter('cases', (c) => c.owner_id === owner_id)
    .sort((a, b) => new Date(b.created_at) - new Date(a.created_at));
  res.status(200).json({ cases });
}

// GET /api/cases/:case_id
function getCase(req, res, body, query, params) {
  const c = store.find('cases', (c) => c.case_id === params.case_id);
  if (!c) return res.status(404).json({ error: 'Case not found' });
  res.status(200).json(c);
}

// POST /api/cases - direct save (e.g. voice-input drafts), mirrors saveCase() in script.js
function createCase(req, res, body) {
  const { owner_id, issue_summary, full_conversation_log, category, language, region } = body;
  const owner = owner_id || req.user?.user_id;
  if (!owner) return res.status(400).json({ error: 'owner_id is required' });
  if (!issue_summary) return res.status(400).json({ error: 'issue_summary is required' });

  const case_id = `case_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;
  const created_at = new Date().toISOString();

  store.insert('cases', {
    case_id,
    owner_id: owner,
    issue_summary: issue_summary.slice(0, 120),
    category: category || null,
    language: language || 'en',
    full_conversation_log: full_conversation_log || [],
    guidance_response: null,
    verdict: null,
    notice_generated: false,
    region: region || null,
    created_at,
  });

  res.status(201).json({ case_id, owner_id: owner, timestamp: created_at });
}

module.exports = { listCases, getCase, createCase };

File: rightsmitra-backend/routes/cases.py
"""Case creation and retrieval endpoints."""

from __future__ import annotations

from datetime import datetime, timezone

from flask import Blueprint, jsonify, request

from routes.helpers import current_user, generate_id, request_json
from storage import filter_rows, find, insert


cases = Blueprint("cases", __name__)


@cases.get("/api/cases")
def list_cases():
    user = current_user()
    owner_id = request.args.get("owner_id") or (user or {}).get("user_id")
    if not owner_id:
        return jsonify(error="owner_id query param required (guest_id or user_id)"), 400
    rows = filter_rows("cases", lambda case: case.get("owner_id") == owner_id)
    rows.sort(key=lambda case: case.get("created_at", ""), reverse=True)
    return jsonify(cases=rows)


@cases.get("/api/cases/<case_id>")
def get_case(case_id: str):
    case = find("cases", lambda item: item.get("case_id") == case_id)
    return (jsonify(case), 200) if case else (jsonify(error="Case not found"), 404)


@cases.post("/api/cases")
def create_case():
    body = request_json()
    user = current_user()
    owner_id = body.get("owner_id") or (user or {}).get("user_id")
    issue_summary = body.get("issue_summary")
    if not owner_id:
        return jsonify(error="owner_id is required"), 400
    if not isinstance(issue_summary, str) or not issue_summary:
        return jsonify(error="issue_summary is required"), 400

    created_at = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    case_id = generate_id("case")
    insert(
        "cases",
        {
            "case_id": case_id,
            "owner_id": owner_id,
            "issue_summary": issue_summary[:120],
            "category": body.get("category"),
            "language": body.get("language") or "en",
            "full_conversation_log": body.get("full_conversation_log") or [],
            "guidance_response": None,
            "verdict": None,
            "notice_generated": False,
            "region": body.get("region"),
            "created_at": created_at,
        },
    )
    return jsonify(case_id=case_id, owner_id=owner_id, timestamp=created_at), 201
